### PR TITLE
PWX-29219: Use cache for getting pod list in health monitor

### PR DESCRIPTION
**Pull request by pp511**
**What type of PR is this?**
>improvement

**What this PR does / why we need it**:
https://portworx.atlassian.net/browse/PWX-29219 
Use Cache interface to get Pod list in health monitor

**Does this PR change a user-facing CRD or CLI?**:
no

**Is a release note needed?**:
yes
```release-note - Improvement
Use Cache to pull Pod list in health monitor to optimize the total stork memory consumption.
```

**Does this change need to be cherry-picked to a release branch?**:
yes. 23.3

**Notes**
- Updated the existing cache to using controllercache.Cache instead of manager client.
- Using controllercache.TransformByObject to modify Pod and only store information that is necessary. We can expand it to use other Params if needed later.
- healthmonitor [https://jenkins.pwx.dev.purestorage.com/job/Stork/view/Stork%2023.2-dev%20%20C7/job/23-2-dev-healthmonitor-k8s125-0/366/] and migration jobs [https://jenkins.pwx.dev.purestorage.com/job/Stork/view/Stork%2023.2-dev%20%20C7/job/23-2-migration-k8s-1-24-0/394/parameters/] to make sure we don't have any regressions. 
- Either need to come up with a fake cache or design UTs specifically for cache. For failing UTs, I had to add  !reflect.ValueOf(storkcache.Instance()).IsNil() so we don't use the cache insterfaces.

### CODE REVIEW

File: pkg/monitor/monitor.go [hunk 1]
```diff
@@ -241,7 +242,8 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
 	if err == nil {
 		return
 	}
-	pods, err := core.Instance().GetPods("", nil)
+
+	pods, err := storkcache.Instance().ListPods()
```
Comment (diptiranjanpx) [inline pkg/monitor/monitor.go:246]: Is it ideal to list Pods from cache where pods are the objects which are most likely change?

Comment (pp511) [reply pkg/monitor/monitor.go]: Cache is always eventually consistent. However in this case we need to optimize the total space we are using to store Pod lists.

File: pkg/cache/cache.go [hunk 2]
```diff
@@ -30,26 +32,67 @@ type SharedInformerCache interface {
...
+	transformMap := controllercache.TransformByObject{
+		&corev1.Pod{}: func(obj interface{}) (interface{}, error) {
+			podResource, ok := obj.(*corev1.Pod)
+			if !ok {
+				return nil, fmt.Errorf("unexpected object type: %T", obj)
+			}
+			currPod := corev1.Pod{}
+			currPod.Name = podResource.Name
+			currPod.Namespace = podResource.Namespace
+			currPod.Namespace = podResource.Namespace
```
Comment (pp511) [inline pkg/cache/cache.go:70]: noticed that Namespace is being assigned twice. Will update along with other review comments..if any.

File: pkg/cache/cache.go [hunk 3]
```diff
@@ -30,26 +32,67 @@ type SharedInformerCache interface {
...
+
+			currPod.Spec.Volumes = podResource.Spec.Volumes
+			currPod.Spec.Containers = podResource.Spec.Containers
```
Comment (adityadani) [inline pkg/cache/cache.go:73]: Do we need Containers?

File: pkg/monitor/monitor.go [hunk 4]
```diff
@@ -241,9 +243,17 @@ func (m *Monitor) cleanupDriverNodePods(node *volume.NodeInfo) {
 	if err == nil {
 		return
 	}
-	pods, err := core.Instance().GetPods("", nil)
+
+	var pods *v1.PodList
+	if !reflect.ValueOf(storkcache.Instance()).IsNil() {
```
Comment (adityadani) [inline pkg/monitor/monitor.go:248]: nit: if storkcache.Instance() != nil should work as well

**Review by adityadani — APPROVED** [review]
looks good

File: pkg/cache/cache.go (hunk 3 above)
Comment (pp511) [reply pkg/cache/cache.go:72]: We do not need it in health-monitor which is the only place where we are using this Cache for listing Pods. However, we are using  Containers in webhook and might later want to update that interface to use the cache instead. So I will keep this for now.

File: pkg/monitor/monitor.go (hunk 4 above)
Comment (pp511) [reply pkg/monitor/monitor.go]: done

File: pkg/cache/cache.go [hunk 5]
```diff
@@ -30,26 +32,67 @@ type SharedInformerCache interface {
 
 	// ListApplicationRegistrations lists the application registration CRs from the cache
 	ListApplicationRegistrations() (*storkv1alpha1.ApplicationRegistrationList, error)
+
+	// ListPods lists the all the Pods from the cache
```
[the remark this answers is not in the document]
Comment (pp511) [reply pkg/cache/cache.go]: done

File: pkg/cache/cache.go (hunk 2 above)
Comment (pp511) [reply pkg/cache/cache.go]: done

**Review by adityadani — APPROVED** [review]

**Pull request merged**